project/src/js/ui/clock/clockCtrl.js

```diff
@@ -0,0 +1,518 @@
+import settings from '../../settings';
+import m from 'mithril';
+import clockSettings from './clockSettings';
+
+export default function controller() {
+  let clockMap = null;
+  const isRunning = m.prop(false);
+  const clockObj = m.prop();
+
+  function reload() {
+    clockMap = {
+      'simple': simpleClock.bind(undefined, Number(settings.clock.simple.time()) * 60, m.redraw),
+      'increment': incrementClock.bind(undefined, Number(settings.clock.increment.time()) * 60, Number(settings.clock.increment.increment()), m.redraw),
+      'delay': delayClock.bind(undefined, Number(settings.clock.delay.time()) * 60, Number(settings.clock.delay.increment()), m.redraw),
+      'bronstein': bronsteinClock.bind(undefined, Number(settings.clock.bronstein.time()) * 60, Number(settings.clock.bronstein.increment()), m.redraw),
+      'hourglass': hourglassClock.bind(undefined, Number(settings.clock.hourglass.time()) * 60, m.redraw),
+      'stage': stageClock.bind(undefined, settings.clock.stage.stages(), Number(settings.clock.stage.increment()), m.redraw)
+    };
+    clockObj(clockMap[settings.clock.clockType()]());
+  }
+
+  reload();
+
+  const clockSettingsCtrl = clockSettings.controller(reload);
+
+  function clockTap (side) {
+    clockObj().clockHit(side);
+  }
+
+  function startStop () {
+    clockObj().startStop();
+  }
+
+  window.plugins.insomnia.keepAwake();
+
+  return {
+    isRunning,
+    startStop,
+    clockSettingsCtrl,
+    clockObj,
+    reload,
+    clockTap,
+    onunload: () => {
+      window.plugins.insomnia.allowSleepAgain();
+      window.StatusBar.show();
+      if (clockObj().clockInterval) {
+        clearInterval(clockObj().clockInterval);
+      }
+    }
+  };
+}
+
+function simpleClock(time, draw) {
```

Comment:
My idea here was to keep the clock code relatively functional by using
callbacks rather than calling functions from upper scope (e.g.,
"m.redraw()") so that we can more easily use them generically in other
contexts (e.g., timed OTB games). Don't feel strongly though and we can
always revert if useful so will make this change.


On Tue, Jun 21, 2016 at 9:53 AM, Vincent Velociter <notifications@github.com
> wrote:

> In project/src/js/ui/clock/clockCtrl.js
> <https://github.com/veloce/lichobile/pull/354#discussion_r67871909>:
>
> > +    startStop,
> > +    clockSettingsCtrl,
> > +    clockObj,
> > +    reload,
> > +    clockTap,
> > +    onunload: () => {
> > +      window.plugins.insomnia.allowSleepAgain();
> > +      window.StatusBar.show();
> > +      if (clockObj().clockInterval) {
> > +        clearInterval(clockObj().clockInterval);
> > +      }
> > +    }
> > +  };
> > +}
> > +
> > +function simpleClock(time, draw) {
>
> I think the name draw here (and every other occurence below) is ambiguous
> because it can be confused with a game draw, and in the whole app it is
> named redraw.
>
> Actually the parameter is unnecessary, every draw() can be rewritten as
> m.redraw(). It would clarify things imo.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/veloce/lichobile/pull/354/files/47416a130f9a4a487049766a55dcb36b301e187e#r67871909>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA6Rdvs9VoN7kVo4gy7vMVIA9rfGuO0Rks5qN-zOgaJpZM4I06tl>
> .
>